go/stats/prombackend/collectors.go

```diff
@@ -0,0 +1,214 @@
+func (c *multiCountersCollector) Collect(ch chan<- prom.Metric) {
+	for mc, desc := range c.multiCounters {
+		for lvs, val := range mc.Counters.Counts() {
+			labelValues := strings.Split(lvs, ".")
+			value := float64(val)
+			ch <- prom.MustNewConstMetric(desc, prom.CounterValue, value, labelValues...)
+		}
+	}
+}
+
+type multiGaugesCollector struct {
+	multiGauges map[*stats.MultiGauges]*prom.Desc
+}
+
+// Describe implements Collector.
+func (c *multiGaugesCollector) Describe(ch chan<- *prom.Desc) {
+	for _, desc := range c.multiGauges {
+		ch <- desc
+	}
+}
+
+// Collect implements Collector.
+func (c *multiGaugesCollector) Collect(ch chan<- prom.Metric) {
+	for mc, desc := range c.multiGauges {
+		for lvs, val := range mc.Gauges.Counters.Counts() {
+			labelValues := strings.Split(lvs, ".")
+			value := float64(val)
+			ch <- prom.MustNewConstMetric(desc, prom.GaugeValue, value, labelValues...)
+		}
+	}
+}
+
+type multiCountersFuncCollector struct {
+	multiCountersFunc map[*stats.MultiCountersFunc]*prom.Desc
+}
+
+// Describe implements Collector.
+func (c *multiCountersFuncCollector) Describe(ch chan<- *prom.Desc) {
+	for _, desc := range c.multiCountersFunc {
+		ch <- desc
+	}
+}
+
+// Collect implements Collector.
+func (c *multiCountersFuncCollector) Collect(ch chan<- prom.Metric) {
+	for mcf, desc := range c.multiCountersFunc {
+		for lvs, val := range mcf.CountersFunc.Counts() {
+			labelValues := strings.Split(lvs, ".")
+			value := float64(val)
+			ch <- prom.MustNewConstMetric(desc, prom.CounterValue, value, labelValues...)
+		}
+	}
+}
+
+type metricCollector struct {
+	m  map[*stats.Counter]*prom.Desc
+	vt stats.ValueType
+}
+
+// Describe implements Collector.
+func (c *metricCollector) Describe(ch chan<- *prom.Desc) {
+	for _, desc := range c.m {
+		ch <- desc
+	}
+}
+
+// Collect implements Collector.
+func (c *metricCollector) Collect(ch chan<- prom.Metric) {
+	for i, desc := range c.m {
+		val := i.Get()
+		if i.Get() == 0 {
+			val = 1
+		}
+		ch <- prom.MustNewConstMetric(desc, toPromValueType(c.vt), float64(val))
+	}
+}
+
+type timingsCollector struct {
+	timings map[*stats.Timings]*prom.Desc
+}
+
+// Describe implements Collector.
+func (c *timingsCollector) Describe(ch chan<- *prom.Desc) {
+	for _, desc := range c.timings {
+		ch <- desc
+	}
+}
+
+// Collect implements Collector.
+func (c *timingsCollector) Collect(ch chan<- prom.Metric) {
+	for t, desc := range c.timings {
+		for cat, his := range t.Histograms() {
+			ch <- prom.MustNewConstHistogram(
+				desc,
+				uint64(his.Count()),
+				float64(his.Total()),
+				makePromBucket(his.Cutoffs(), his.Buckets()),
+				cat)
+		}
+	}
+}
+
+func makePromBucket(cutoffs []int64, buckets []int64) map[float64]uint64 {
+	output := make(map[float64]uint64)
+	last := uint64(0)
+	for i := range cutoffs {
+		key := float64(cutoffs[i]) / 1000000000
+		//TODO(zmagg): int64 => uint64 conversion. error if it overflows?
+		output[key] = uint64(buckets[i]) + last
```

Comment:
It seems to me the buckets should just be uint64 from the start.

Reply:
Hmm. They're `sync2.AtomicInt64`s in `histogram.go`.

Reply:
I think that's wrong too :)

Of course the "right" thing seems to me that we add a `sync2.AtomicUint64` and change all the histogram metrics to use that so we don't have to do this dance here.

@sougou any opinion here?